holidays/countries/paraguay.py

```diff
@@ -27,82 +39,201 @@ class Paraguay(HolidayBase):
+            # public holidays in Paraguay on account
+            # of the upcoming visit of Pope Francis in Paraguay
+            (JUL, 10, "Public Holiday"),
+            (JUL, 11, "Public Holiday"),
+            # 2 year-end public sector holidays
+            (DEC, 24, "Public sector Holiday"),
+            (DEC, 31, "Public sector Holiday"),
+        ),
+        2016: (
+            # public sector holiday to let civil servants
+            # begin their Holy Week earlier
+            (MAR, 23, "Public sector Holiday"),
+            (DEC, 26, "Public Holiday"),
+        ),
+        2017: (
+            (JAN, 2, "Public Holiday"),
+            # public sector holiday to let civil servants
+            # begin their Holy Week earlier
+            (MAR, 28, "Public sector Holiday"),
+        ),
+        2018: (
+            # 2 year-end public sector holidays
+            (DEC, 24, "Public sector Holiday"),
+            (DEC, 31, "Public sector Holiday"),
+        ),
+        2019: (
+            # public sector holiday to let civil servants
+            # begin their Holy Week earlier
+            (APR, 17, "Public sector Holiday"),
+            # 2 year-end public sector holidays
+            (DEC, 24, "Public sector Holiday"),
+            (DEC, 31, "Public sector Holiday"),
+        ),
+        2020: (
+            # public sector holiday to let civil servants
+            # begin their Holy Week earlier
+            (APR, 8, "Public sector Holiday"),
+        ),
+        2021: (
+            # 2 year-end public sector holidays
+            (DEC, 24, "Public sector Holiday"),
+            (DEC, 31, "Public sector Holiday"),
+        ),
+        2022: (
+            # public sector holiday to let civil servants
+            # begin their Holy Week earlier
+            (APR, 13, "Public sector Holiday"),
+            # public sector holiday due to the annual May 1st
+            # public holiday falling on a Sunday
+            (MAY, 2, "Public Holiday"),
+        ),
+    }
+
+    def _add_holiday(self, dt: date, name: str) -> None:
+        if self.observed or dt.weekday() not in WEEKEND:
+            self[dt] = name
 
     def _populate(self, year):
         super()._populate(year)
 
         # New Year's Day
-        if not self.observed and date(year, JAN, 1).weekday() in WEEKEND:
-            pass
-        else:
-            self[date(year, JAN, 1)] = "Año Nuevo [New Year's Day]"
+        self._add_holiday(date(year, JAN, 1), "Año Nuevo [New Year's Day]")
 
         # Patriots day
-        name = "Día de los Héroes de la Patria [Patriots Day]"
-
-        if not self.observed and date(year, MAR, 1).weekday() in WEEKEND:
-            pass
-        elif date(year, MAR, 1).weekday() >= WED:
-            self[date(year, MAR, 1) + rd(weekday=MO(+1))] = name
+        dates_obs = {
+            2013: (MAR, 4),
+            2016: (FEB, 29),
+            2018: (FEB, 26),
+            2022: (FEB, 28),
+        }
+        if year in dates_obs:
```

Comment:
Optional (here and below): 
```
self[date(year, *dates_obs.get(year, (MAR, 1)))] = "Día de los Héroes de la Patria [Patriots Day]"
```

Reply:
Thank you! That's why I like Python, for things like this :)